lib/bb_statements/api_client.rb

```diff
@@ -0,0 +1,127 @@
+require 'date'
+require 'json'
+require 'logger'
+require 'tempfile'
+require 'faraday'
+require 'uri'
+
+module BancoBrasilStatements
+  class ApiClient
+    TOKEN_EXPIRE_TIME = 570
+
+    attr_accessor :config, :default_headers, :access_token_requested_at
+
+    # Initializes the ApiClient
+    # @option config [Configuration] Configuration for initializing the object, default to Configuration.default
+    def initialize(config = Configuration.default)
+      @config = config
+      @user_agent = "bb-statements-gem/#{VERSION}/ruby"
+      @default_headers = {
+        'Content-Type' => 'application/json',
+        'User-Agent' => @user_agent
+      }
+      @access_token_requested_at = nil
+    end
+
+    def self.default
+      @@default ||= ApiClient.new
+    end
+
+    def call_api(http_method, path, opts = {})
+      request = build_request(path, opts)
+      response = request.send(http_method.downcase.to_sym)
+
+      if @config.debugging
+        @config.logger.debug "HTTP response body ~BEGIN~\n#{response.body.force_encoding('utf-8')}\n~END~\n"
+      end
+
+      unless response.success?
+        fail BancoBrasilStatements::ApiError.new(:code => response.status,
+                                                 :response_headers => response.headers,
+                                                 :response_body => response.body.force_encoding('utf-8')),
+                                                 response.reason_phrase
+      end
+
+      JSON.parse(response.body)
+    rescue Faraday::TimeoutError => e
+      fail BancoBrasilStatements::ApiError.new('Connection timed out')
+    rescue Faraday::ConnectionFailed => e
+      fail BancoBrasilStatements::ApiError.new('Connection failed')
+    rescue Faraday::RequestTimeoutError => e
+      fail BancoBrasilStatements::ApiError.new('Connection timed out')
+    end
```

Comment:
```suggestion
    rescue Faraday::Error => e
      fail BancoBrasilStatements::ApiError.new('Connection timed out', cause: e)
    end
```

https://www.honeybadger.io/blog/nested-errors-in-ruby-with-exception-cause/

https://ruby-doc.org/core-2.6/Exception.html